text/shared/src/main/scala/fs2/data/text/render/internal/StreamPrinter.scala

```diff
@@ -21,246 +21,275 @@ import cats.collections.Dequeue
-        evt match {
-          case Annotated.Text(text, _)         => ((fit, hpl, lines, col + text.size), Some(text))
-          case Annotated.Line(pos) if fit == 0 => ((fit, pos + width, lines, lines.head.size), Some("\n" + lines.head))
-          case Annotated.Line(_)               => ((fit, hpl, lines, col + 1), Some(" "))
-          case Annotated.LineBreak(pos) if fit == 0 =>
-            ((fit, pos + width, lines, lines.head.size), Some("\n" + lines.head))
-          case Annotated.LineBreak(_)                            => (acc, None)
-          case Annotated.GroupBegin(Position.TooFar) if fit == 0 => ((0, hpl, lines, col), None)
-          case Annotated.GroupBegin(Position.Small(pos)) if fit == 0 =>
-            ((if (pos <= hpl) 1 else 0, hpl, lines, col), None)
-          case Annotated.GroupBegin(_)           => ((fit + 1, hpl, lines, col), None)
-          case Annotated.GroupEnd(_) if fit == 0 => (acc, None)
-          case Annotated.GroupEnd(_)             => ((fit - 1, hpl, lines, col), None)
-          case Annotated.IndentBegin(_) =>
-            ((fit, hpl, NonEmptyList(lines.head + (" " * indentSize), lines.tail), col), None)
-          case Annotated.IndentEnd(_) =>
-            ((fit, hpl, NonEmptyList(lines.head.drop(indentSize), lines.tail), col), None)
-          case Annotated.AlignBegin(_) =>
-            ((fit, hpl, (" " * col) :: lines, col), None)
-          case Annotated.AlignEnd(_) =>
-            ((fit, hpl, NonEmptyList.fromList(lines.tail).getOrElse(NonEmptyList.one("")), col), None)
-        }
+  // rendering
+
+  private def renderText(text: String, ctx: RenderingContext, chunkAcc: StringBuilder): Unit = {
+    ctx.col += text.size
+    chunkAcc.append(text)
+  }
+
+  private def renderLine(pos: Int, ctx: RenderingContext, chunkAcc: StringBuilder): Unit = if (ctx.fit == 0) {
+    ctx.hpl = pos + width
+    ctx.col = ctx.lines.head.size
+    val _ = chunkAcc.append('\n').append(ctx.lines.head)
+  } else {
+    ctx.col += 1
+    chunkAcc.append(' ')
+  }
+
+  private def renderLineBreak(pos: Int, ctx: RenderingContext, chunkAcc: StringBuilder): Unit =
+    if (ctx.fit == 0) {
+      ctx.hpl = pos + width
+      ctx.col = ctx.lines.head.size
+      val _ = chunkAcc.append('\n').append(ctx.lines.head)
+    }
 
+  private def renderGroupBegin(pos: Position, ctx: RenderingContext): Unit =
+    if (ctx.fit == 0) {
+      pos match {
+        case Position.TooFar =>
+        // too far, do nothing
+        case Position.Small(pos) =>
+          ctx.fit = if (pos <= ctx.hpl) 1 else 0
       }
-      .map(_._2)
-      .unNone
+    } else {
+      ctx.fit += 1
+    }
+
+  private def renderGroupEnd(ctx: RenderingContext): Unit =
+    if (ctx.fit > 0) {
+      ctx.fit -= 1
+    }
+
+  private def renderIndentBegin(ctx: RenderingContext): Unit = {
+    ctx.lines = NonEmptyList(ctx.lines.head + (" " * indentSize), ctx.lines.tail)
```

Comment:
Do you think a `StringBuilder` that is allocated to the right capacity could help speeding up here? Like first append the head, then `indentSize` spaces?

Reply:
I was think it would actually be way better to cache the padding. In a structured data rendering the same indent size will be found over and over again. I will try to reuse the padding that was already encountered for a indent depth, rather than this.

Reply:
I don't have much time right now to fine tune this. I will merge as is, since this fixes a big regression and publish a new bugfix release.